src/lib/builder.rs

```diff
@@ -0,0 +1,171 @@
+///   -> Result<Node<TokId>,
+///            (Option<Node<TokId>>, Vec<ParseError<TokId>>)>
+/// ```
+pub fn process_file<'a, TokId, P, Q>(inp: P,
+                                     outp: Q)
+                                  -> Result<(HashMap<String, TokId>), Box<Error>>
+                               where TokId: Copy + Debug + Eq + TryFrom<usize> + TypeName,
+                                     P: AsRef<Path>,
+                                     Q: AsRef<Path>
+{
+    let inc = read_to_string(&inp).unwrap();
+
+    let grm = match yacc_grm(YaccKind::Eco, &inc) {
+        Ok(x) => x,
+        Err(s) => {
+            panic!("{:?}", s);
+        }
+    };
+    let rule_ids = grm.terms_map().iter()
+                                  .map(|(&n, &i)| (n.to_owned(),
+                                                   TokId::try_from(usize::from(i))
+                                                         .unwrap_or_else(|_| panic!("woo"))))
+                                  .collect::<HashMap<_, _>>();
+
+    let (sgraph, stable) = match from_yacc(&grm, Minimiser::Pager) {
+        Ok(x) => x,
+        Err(s) => {
+            panic!("{:?}", s);
+        }
+    };
+
+    let mut outs = String::new();
+    // Header
+    let mod_name = inp.as_ref().file_stem().unwrap().to_str().unwrap();
+    outs.push_str(&format!("mod {}_y {{", mod_name));
+    outs.push_str(&format!("use lrpar::{{Node, parse_rcvry, ParseError, reconstitute, RecoveryKind}};
+use lrlex::Lexeme;
+
+pub fn parse(lexemes: &Vec<Lexeme<{tn}>>)
+          -> Result<Node<{tn}>, (Option<Node<{tn}>>, Vec<ParseError<{tn}>>)>
+{{", tn=TokId::type_name()));
+
+    // grm, sgraph, stable
+    let mut grm_buf = Vec::new();
+    grm.serialize(&mut Serializer::new(&mut grm_buf)).unwrap();
+    let mut sgraph_buf = Vec::new();
+    sgraph.serialize(&mut Serializer::new(&mut sgraph_buf)).unwrap();
+    let mut stable_buf = Vec::new();
+    stable.serialize(&mut Serializer::new(&mut stable_buf)).unwrap();
+    outs.push_str(&format!("
+    let (grm, sgraph, stable) = reconstitute(&vec!{:?}, &vec!{:?}, &vec!{:?});
+    parse_rcvry(RecoveryKind::MF, &grm, |_| 1, &sgraph, &stable, lexemes)
+", grm_buf, sgraph_buf, stable_buf));
+
+    outs.push_str("}");
+
+    // Footer
+    outs.push_str("}");
+    // If the file we're about to write out already exists with the same contents, then we don't
+    // overwrite it (since that will force a recompile of the file, and relinking of the binary
+    // etc).
+    if let Ok(curs) = read_to_string(&outp) {
+        if curs == outs {
+            return Ok(rule_ids);
+        }
+    }
+    let mut f = File::create(outp)?;
+    f.write_all(outs.as_bytes())?;
+    Ok(rule_ids)
+}
+
+/// This function is called by generated files; it exists so that generated files don't require a
+/// dependency on serde and rmps.
+#[doc(hidden)]
+pub fn reconstitute(grm_buf: &Vec<u8>, sgraph_buf: &Vec<u8>, stable_buf: &Vec<u8>)
```

Comment:
Instead of `&Vec<u8>` here you can say `&[u8]` which uses one less reference.

Reply:
It doesn't save a reference, but it does make things shorter, so I agree: it's better as you suggest.